### PR TITLE
Make as_finite_diff a method, new differentiating func

### LINKED ISSUE #11007: as_finite_diff does not support partial derivatives
partial derivatives is not well supported by `as_finite_diff`:
```
>>> f = Function('f')(x, y)
>>> d2fdxdy = f.diff(x).diff(y)
>>> as_finite_diff(d2fdxdy, wrt=x)
-f(x - 1/2, y) + f(x + 1/2, y)
```
which is incorrect, the correct output would be:
```
  ∂                   ∂                
- ──(f(x - 1/2, y)) + ──(f(x + 1/2, y))
  ∂y                  ∂y
```

related:
https://groups.google.com/forum/#!searchin/sympy/Mathias$20Louboutin/sympy/MIlkgmts8xw/jQ_-EUrUAAAJ
https://groups.google.com/forum/#!searchin/sympy/Mathias$20Louboutin/sympy/mzr_ET2Pu2E/0gK6jxcUDQAJ

a nice addition would be to allow the syntax `wrt=(x, y)`

### PR DESCRIPTION
**Pull request by bjodah**
This pull-request fixes #11007.
It is the result of the discussions in the alternative PRs: #11242 #11177 

In summary this PR does the following:

```
- Deprecates `as_finite_diff`
- New function: `differentiate_finite`
- New method: `Derivative.as_finite_difference`
```

Pinging @parsoyaarihant, @mloubout and @asmeurer for review/feedback.

### CODE REVIEW

**Comment by bjodah** [comment]
Pinging @derDavidT who has helped getting the documentation correct.

File: sympy/core/function.py [hunk 1]
```diff
@@ -1367,6 +1367,86 @@ def _sage_(self):
...
+
+        The algorithm is not restricted to use equidistant spacing, nor
+        do we need to make the approximation around ``x0``, but we can get
+        an expression estimating the derivative at an offset:
+
+        >>> e, sq2 = exp(1), sqrt(2)
+        >>> xl = [x-h, x+h, x+e*h]
+        >>> f(x).diff(x, 1).as_finite_difference(xl, x+h*sq2)
+        2*h*((h + sqrt(2)*h)/(2*h) - (-sqrt(2)*h + h)/(2*h))*f(E*h + x)/\
+((-h + E*h)*(h + E*h)) + (-(-sqrt(2)*h + h)/(2*h) - \
```
Comment (asmeurer) [inline sympy/core/function.py]: These lines should be indented.

File: sympy/calculus/finite_diff.py [hunk 2]
```diff
@@ -382,10 +395,73 @@ def as_finite_diff(derivative, points=1, x0=None, wrt=None):
...
     if len(points) < order+1:
         raise ValueError("Too few points for order %d" % order)
     return apply_finite_diff(order, points, [
-        derivative.expr.subs({wrt: x}) for x in points], x0)
+        Derivative(derivative.expr.subs({wrt: x}), *others) for
+        x in points], x0)
+
+
+as_finite_diff = deprecated(useinstead="Derivative.as_finite_difference",
+                            deprecated_since_version="1.1")(_as_finite_diff)
```
Comment (asmeurer) [inline sympy/calculus/finite_diff.py]: Deprecated things also need a deprecation issue. See https://github.com/sympy/sympy/wiki/Deprecating-policy

**Comment by bjodah** [comment]
Thanks for the link to the wiki page.

Regarding the indentation in the docstring, I couldn't make the doctests pass with `+NORMALIZE_WHITESPACE` (see e.g. [this](http://stackoverflow.com/questions/13395610/wrapping-python-doctest-results-that-are-longer-than-80-characters)) I went with the use of an ellipsis instead.

**Comment by bjodah** [comment]
@mloubout - sorry for the double ping, but I saw that you referenced this PR in https://github.com/opesci/notebooks/issues/11

I am just interested to know if you find this PR helpful for your work involving finite differences?
I am hoping that I managed to address the points you raised in the referenced issues/pull requests, but perhaps there are more things I could improve here?

**Comment by mloubout** [comment]
@bjodah We ended up writing our own finite-difference class as we need a lot of multi-level finite differences operators with proper adjoints. This is however a very helpful improvement of the functionality that will help a lot.

**Comment by bjodah** [comment]
@mloubout thank you for the input.
@asmeurer does the deprecation look alright now?

If there are no further comments I'll merge this in a week or so.

**Pull request merged**